// @nextui-org/navbar/dist/navbar.js
"use client";
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/navbar.tsx
var navbar_exports = {};
__export(navbar_exports, {
  default: () => navbar_default
});
module.exports = __toCommonJS(navbar_exports);
var import_system3 = require("@nextui-org/system");
var import_react_utils4 = require("@nextui-org/react-utils");
var import_framer_motion2 = require("framer-motion");
var import_utils4 = require("@react-aria/utils");

// src/navbar-transitions.ts
var import_framer_transitions = require("@nextui-org/framer-transitions");
var hideOnScrollVariants = {
  visible: {
    y: 0,
    transition: {
      ease: import_framer_transitions.TRANSITION_EASINGS.easeOut
    }
  },
  hidden: {
    y: "-100%",
    transition: {
      ease: import_framer_transitions.TRANSITION_EASINGS.easeIn
    }
  }
};

// src/use-navbar.ts
var import_system = require("@nextui-org/system");
var import_theme = require("@nextui-org/theme");
var import_react_utils = require("@nextui-org/react-utils");
var import_shared_utils = require("@nextui-org/shared-utils");
var import_react = require("react");
var import_utils = require("@react-aria/utils");
var import_use_scroll_position = require("@nextui-org/use-scroll-position");
var import_utils2 = require("@react-stately/utils");
function useNavbar(originalProps) {
  var _a;
  const [props, variantProps] = (0, import_system.mapPropsVariants)(originalProps, import_theme.navbar.variantKeys);
  const {
    ref,
    as,
    parentRef,
    height = "4rem",
    shouldHideOnScroll = false,
    disableScrollHandler = false,
    onScrollPositionChange,
    isMenuOpen: isMenuOpenProp,
    isMenuDefaultOpen,
    onMenuOpenChange = () => {
    },
    motionProps,
    className,
    classNames,
    ...otherProps
  } = props;
  const Component = as || "nav";
  const domRef = (0, import_react_utils.useDOMRef)(ref);
  const prevWidth = (0, import_react.useRef)(0);
  const navHeight = (0, import_react.useRef)(0);
  const [isHidden, setIsHidden] = (0, import_react.useState)(false);
  const handleMenuOpenChange = (0, import_react.useCallback)(
    (isOpen) => {
      onMenuOpenChange(isOpen || false);
    },
    [onMenuOpenChange]
  );
  const [isMenuOpen, setIsMenuOpen] = (0, import_utils2.useControlledState)(
    isMenuOpenProp,
    isMenuDefaultOpen,
    handleMenuOpenChange
  );
  const updateWidth = () => {
    if (domRef.current) {
      const width = domRef.current.offsetWidth;
      if (width !== prevWidth.current) {
        prevWidth.current = width;
      }
    }
  };
  (0, import_utils.useResizeObserver)({
    ref: domRef,
    onResize: () => {
      var _a2;
      const currentWidth = (_a2 = domRef.current) == null ? void 0 : _a2.offsetWidth;
      if (currentWidth !== prevWidth.current) {
        updateWidth();
        setIsMenuOpen(false);
      }
    }
  });
  (0, import_react.useEffect)(() => {
    var _a2;
    updateWidth();
    navHeight.current = ((_a2 = domRef.current) == null ? void 0 : _a2.offsetHeight) || 0;
  }, []);
  const slots = (0, import_react.useMemo)(
    () => (0, import_theme.navbar)({
      ...variantProps,
      hideOnScroll: shouldHideOnScroll
    }),
    [...Object.values(variantProps), shouldHideOnScroll]
  );
  const baseStyles = (0, import_shared_utils.clsx)(classNames == null ? void 0 : classNames.base, className);
  (0, import_use_scroll_position.useScrollPosition)({
    elementRef: parentRef,
    isEnabled: shouldHideOnScroll || !disableScrollHandler,
    callback: ({ prevPos, currPos }) => {
      onScrollPositionChange == null ? void 0 : onScrollPositionChange(currPos.y);
      if (shouldHideOnScroll) {
        setIsHidden((prev) => {
          const next = currPos.y > prevPos.y && currPos.y > navHeight.current;
          return next !== prev ? next : prev;
        });
      }
    }
  });
  const getBaseProps = (props2 = {}) => ({
    ...(0, import_utils.mergeProps)(otherProps, props2),
    "data-hidden": (0, import_shared_utils.dataAttr)(isHidden),
    "data-menu-open": (0, import_shared_utils.dataAttr)(isMenuOpen),
    ref: domRef,
    className: slots.base({ class: (0, import_shared_utils.clsx)(baseStyles, props2 == null ? void 0 : props2.className) }),
    style: {
      "--navbar-height": height,
      ...otherProps == null ? void 0 : otherProps.style,
      ...props2 == null ? void 0 : props2.style
    }
  });
  const getWrapperProps = (props2 = {}) => ({
    ...props2,
    "data-menu-open": (0, import_shared_utils.dataAttr)(isMenuOpen),
    className: slots.wrapper({ class: (0, import_shared_utils.clsx)(classNames == null ? void 0 : classNames.wrapper, props2 == null ? void 0 : props2.className) })
  });
  return {
    Component,
    slots,
    domRef,
    height,
    isHidden,
    disableAnimation: (_a = originalProps.disableAnimation) != null ? _a : false,
    shouldHideOnScroll,
    isMenuOpen,
    classNames,
    setIsMenuOpen,
    motionProps,
    getBaseProps,
    getWrapperProps
  };
}

// src/navbar-context.ts
var import_react_utils2 = require("@nextui-org/react-utils");
var [NavbarProvider, useNavbarContext] = (0, import_react_utils2.createContext)({
  name: "NavbarContext",
  strict: true,
  errorMessage: "useNavbarContext: `context` is undefined. Seems you forgot to wrap component within <Navbar />"
});

// src/navbar-menu.tsx
var import_system2 = require("@nextui-org/system");
var import_react_utils3 = require("@nextui-org/react-utils");
var import_shared_utils2 = require("@nextui-org/shared-utils");
var import_framer_motion = require("framer-motion");
var import_utils3 = require("@react-aria/utils");
var import_react2 = require("react");
var import_react_remove_scroll = require("react-remove-scroll");
var import_overlays = require("@react-aria/overlays");

// src/navbar-menu-transitions.ts
var menuVariants = {
  enter: {
    height: "calc(100vh - var(--navbar-height) - 1px)",
    transition: {
      duration: 0.3,
      easings: "easeOut"
    }
  },
  exit: {
    height: 0,
    transition: {
      duration: 0.25,
      easings: "easeIn"
    }
  }
};

// src/navbar-menu.tsx
var import_jsx_runtime = require("react/jsx-runtime");
var NavbarMenu = (0, import_system2.forwardRef)((props, ref) => {
  var _a, _b;
  const { className, children, portalContainer, motionProps, style, ...otherProps } = props;
  const domRef = (0, import_react_utils3.useDOMRef)(ref);
  const { slots, isMenuOpen, height, disableAnimation, classNames } = useNavbarContext();
  const styles = (0, import_shared_utils2.clsx)(classNames == null ? void 0 : classNames.menu, className);
  const MenuWrapper = (0, import_react2.useCallback)(
    ({ children: children2 }) => {
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react_remove_scroll.RemoveScroll, { forwardProps: true, enabled: isMenuOpen, removeScrollBar: false, children: children2 });
    },
    [isMenuOpen]
  );
  const contents = disableAnimation ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(MenuWrapper, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "ul",
    {
      ref: domRef,
      className: (_a = slots.menu) == null ? void 0 : _a.call(slots, { class: styles }),
      "data-open": (0, import_shared_utils2.dataAttr)(isMenuOpen),
      style: {
        "--navbar-height": height
      },
      ...otherProps,
      children
    }
  ) }) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_framer_motion.AnimatePresence, { mode: "wait", children: isMenuOpen ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(MenuWrapper, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_framer_motion.motion.ul,
    {
      ref: domRef,
      layoutScroll: true,
      animate: "enter",
      className: (_b = slots.menu) == null ? void 0 : _b.call(slots, { class: styles }),
      "data-open": (0, import_shared_utils2.dataAttr)(isMenuOpen),
      exit: "exit",
      initial: "exit",
      style: {
        "--navbar-height": height,
        ...style
      },
      variants: menuVariants,
      ...(0, import_utils3.mergeProps)(motionProps, otherProps),
      children
    }
  ) }) : null });
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_overlays.Overlay, { portalContainer, children: contents });
});
NavbarMenu.displayName = "NextUI.NavbarMenu";
var navbar_menu_default = NavbarMenu;

// src/navbar.tsx
var import_jsx_runtime2 = require("react/jsx-runtime");
var Navbar = (0, import_system3.forwardRef)((props, ref) => {
  const { children, ...otherProps } = props;
  const context = useNavbar({ ...otherProps, ref });
  const Component = context.Component;
  const [childrenWithoutMenu, menu] = (0, import_react_utils4.pickChildren)(children, navbar_menu_default);
  const content = /* @__PURE__ */ (0, import_jsx_runtime2.jsxs)(import_jsx_runtime2.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime2.jsx)("header", { ...context.getWrapperProps(), children: childrenWithoutMenu }),
    menu
  ] });
  return /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(NavbarProvider, { value: context, children: context.shouldHideOnScroll ? /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(
    import_framer_motion2.motion.nav,
    {
      animate: context.isHidden ? "hidden" : "visible",
      initial: false,
      variants: hideOnScrollVariants,
      ...(0, import_utils4.mergeProps)(context.getBaseProps(), context.motionProps),
      children: content
    }
  ) : /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(Component, { ...context.getBaseProps(), children: content }) });
});
Navbar.displayName = "NextUI.Navbar";
var navbar_default = Navbar;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {});
